_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+---
+layout: posts
+title: Configurable Builds - Part 1
+authors:
+ - gregestren
+---
+
+The larger your project gets, the more likely you have to build it in different
+ways.
+
+Maybe you need Android, iOS, and desktop versions of your app. Maybe you build
+C++ for different platforms. Or maybe you maintain a popular library that
+different users need different features from.
+
+These are examples of *configuration*: the process of building the same *code*
+with different *settings* to customize it for specific needs. In Bazel-speak,
+this means building a
+[target](https://docs.bazel.build/versions/master/build-ref.html#targets) with
+a set of
+[flags](https://docs.bazel.build/versions/master/user-manual.html#flags-options):
+
+```sh
+$ bazel build //my:cc_binary --cpu=x86
+$ bazel build //my:cc_binary --cpu=arm
+$ bazel build //my:android_binary --android_sdk=@androidsdk//:sdk-28
+```
+
+This post reviews Bazel's *configurability work*, which is an ongoing effort to
+to make these tasks simple, flexible, and powerful. This is a deeper dive into
+themes covered in Bazel's [configurability
+roadmap](https://bazel.build/roadmaps/configuration.html).
+
+## History
+
+Bazel's a
+[powerful](https://docs.bazel.build/versions/master/bazel-vision.html) build
+tool that's especially suited for large,
+[multi-language](/2018/12/05/multilanguage-build-system.html) builds. But it
+was born out of Google, which historically wrote code for homogeneous fleets of
+Linux servers with little need for customization. What customization was needed
+was achieved with ad hoc flags and logic built
+[straight](https://source.bazel.build/bazel/+/144912e7b7a86b45e07f79e76f6fed20890acb36:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=258)
```

Comment:
suggestion. make the link cover "logic built straight into the tool"
That links the full concept to the example.

Reply:
Re: "build tool" -> "tool". Done. And hmm, yeah maybe "straight" is unnecessary too. Leaving this conversation open to contemplate, and because I have to rush out in 5 minutes. :)

Re: link. I was trying to express that both the flags and logic associated with them are part of the historic cruft we're trying to clean up. So the full conceptual link in my mind is "ad hoc flags and logic built into the tool", which seems like a huge link. What about just linking "ad hoc flags and logic"?

Reply:
Closing on account of my last changes. Feel free to re-open if desired.